### DISCUSSION TITLE
PixelTypeInfo could use a BitsPerComponent property

### DISCUSSION THREAD

**Discussion by 0xC0000054** [SixLabors/ImageSharp#1394] [Ideas] [answered]
Currently the only method to determine the image bit-depth when using `Identify` is to guess based on the value of the BitsPerPixel property.
The BitsPerPixel property can be ambiguous for some image formats, e.g. `L16` and `La16` have the same number of bits-per-pixel.

A BitsPerComponent property would allow callers to determine the image bit depth without any guesswork.

This would simplify the case where a caller does not know in advance if the image is 8-bits-per-component or 16-bits-per-component.

**Comment by JimBobSquarePants** [comment, marked as answer]
The idea behind the current API is that further information is obtainable upon demand.

For example if you use.

```c#
Image.Identify(stream, out IImageFormat detectedFormat)
```

Then you can use the returned format to determine other format specific metadata. 

```c#
if (detectedFormat is PngFormat pngFormat)
{
    PngMetadata pngMeta = imageInfo.Metadata.GetPngMetadata();

    PngColorType? colorType = pngMeta.ColorType;
}
```
You can then use that returned value to tell you whether the original pixel format contains an alpha component.

**Reply by antonfirsov** [reply]
@JimBobSquarePants this is probably rather a workaround than the solution we want to stick with. I think there are many questions around `PixelTypeInfo` that should be answered.

At the moment it is a chunk, which has two roles:
1. A runtime descriptor for the `TPixel` in `Image<TPixel>`. I consider this a very important feature. (For example querying a `TPixel` for a `.HasAlpha` property can enable optimizations, which will help saving up to 10% on LoadResizeSave.)
2. A descriptor for the native pixel type, when used with `Image.Identify`

Currently we are mixing the two concerns.
- Which actually makes *some* sense, because there is an overlap between them. Most graphics API-s decode images as close to their native format as possible. In case of PNG or BMP it could be a perfect match. The fact that `Image.Load` is always decoding into `Rgba32` is something we'd like to change.
- Is an issue on the other hand, since it's not possible to guarantee this match. For example: we'll probably never introduce a pixel format for `YCbCr` which is the native format of Jpeg. We will keep decoding it to `Rgba32` (later to `Rgb24`)

@0xC0000054 can you elaborate a bit more about your use case?
- A/ What is your application, and how does it use the information about the bit depth?
- B/ Do you have some prior knowledge on the images being fed to `Identify`? (Eg. all of them are PNG-s.)
- C/ What about a case when it's not possible to match the native pixel format?

**Reply by JimBobSquarePants** [reply]
I'm not sure it's accurately described as a workaround. The API is working exactly as intended IMO. 

When using `Identify`, `BitsPerPixel` is calculated from the image headers of the encoded images and is not related to `TPixel`.

When decoding the image via `Decode` the `BitsPerPixel` property is calculated from the `TPixel` value.

If you're decoding to `L16` or `La16` you already know exactly what the layout is of each pixel and I would presume choosing to use each type is an educated. The metadata supplied allows you to make that educated decision.

It could be considered a little confusing but only, as you say, because we do not currently optimize the pixel format when decoding via the non-generic API. 

We've already decided that we should optimize but that requires all the SIMD work to be done first. 

I would definitely like to see a `HasAlpha()` method on the `IPixel` type to provide a bunch of optimizations but I don't think that's relevant in this case.

**Reply by antonfirsov** [reply]
Speaking about the "get some property of the encoded images pixel type info" feature:
I referred to your suggestion as a workaround, because it could be also implemented without the need for expecting the user to dig into format-specific metadata. Theoretically, it could be just a function takimg the stream and returning a feature rich `PixelTypeInfo`.

> I would definitely like to see a HasAlpha() method on the IPixel type

This property is not boumd to individual pixels,  but rather to the **pixel types** of thr image, which is what `PixelTypeInfo` intends to represent.

**Reply by JimBobSquarePants** [reply]
I think it’s more useful on the pixel format so it can be read by any TPixel method.

**Reply by 0xC0000054** [reply]
@antonfirsov 


> A/ What is your application, and how does it use the information about the bit depth?

I use the bit depth to determine if the image should be loaded as `Rgba64` or `Rgba32`.

> B/ Do you have some prior knowledge on the images being fed to Identify? (Eg. all of them are PNG-s.)

No, it could be any format.

> C/ What about a case when it's not possible to match the native pixel format?

Then it will be converted to `Rgba64` or `Rgba32`.

**Reply by antonfirsov** [reply]
@JimBobSquarePants I wouldn't add an extensible (!) list of new properties on `IPixel`, because (1) it's already too heavy. (2) Pixel and Pixel Type/Format are really different domain model terms. I believe the right way of separating these concerns is to distinguish them on type system level. WIC has a similar abstraction:
https://docs.microsoft.com/en-us/windows/win32/api/wincodec/nn-wincodec-iwicpixelformatinfo

> [...] so it can be read by any TPixel method.

A way to get a `PixelTypeInfo` for a generic `TPixel` type argument could be as simple as `default(TPixel).GetPixelTypeInfo()` or `PixelOperations<TPixel>.Instance.GetPixelTypeInfo()`

**Reply by JimBobSquarePants** [reply]
If you don't enforce it as a contract how do you ensure that all `TPixel` implementations return the correct value? Or do you mean add the `GetPixelTypeInfo` method to the interface?

**Reply by antonfirsov** [reply]
Yeah, adding `GetPixelTypeInfo` to `IPixel` would be the safest way so it's not an optional overwrite, we can then encapsulate all pixel format related information in that type.

Because of the overlapping concepts, I didn't see anything against using the same type for this and the stuff returned `IImageInfo.PixelTypeInfo`, so we can provide a format-agnostic API to observe raw pixelformat information in encoded streams. We just need to make  `IImageInfo.PixelTypeInfo` (or maybe individual `PixelTypeInfo` properties?) nullable.

**Reply by JimBobSquarePants** [reply]
Yep, sounds good👌

**Comment by antonfirsov** [comment]
> use the bit depth to determine if the image should be loaded as `Rgba64` or `Rgba32`.

@0xC0000054 what do you do with the image after decoding it? Do you apply built-in processors like Resize, or do you need to access the pixel buffers manually (`image[x,y]`, `image.GetPixelRowSpan(y)`)?

Assuming the first case, if the (non generic) `Image.Load` would decode the (non-generic) `Image` into the right pixel format, it would fix your issue without any extra code.

**Comment by JimBobSquarePants** [comment]
@0xC0000054 V4 will have this information. We'll provide per component bit depth and ordering as well as color type metadata
